fe/src/main/java/org/apache/doris/transaction/GlobalTransactionMgr.java

```diff
@@ -1346,81 +359,71 @@ public TransactionIdGenerator getTransactionIDGenerator() {
+
     }
 
-    public TransactionState getTransactionStateByCallbackIdAndStatus(long callbackId, Set<TransactionStatus> status) {
-        readLock();
+    public TransactionState getTransactionStateByCallbackIdAndStatus(long dbId, long callbackId, Set<TransactionStatus> status) {
         try {
-            for (TransactionState txn : idToTransactionState.values()) {
-                if (txn.getCallbackId() == callbackId && status.contains(txn.getTransactionStatus())) {
-                    return txn;
-                }
-            }
-        } finally {
-            readUnlock();
+            DatabaseTransactionMgr dbTransactionMgr = getDatabaseTransactioMgr(dbId);
+            return dbTransactionMgr.getTransactionStateByCallbackIdAndStatus(callbackId, status);
+        } catch (AnalysisException e) {
+            LOG.warn("Get transaction by callbackId and status failed", e);
+            return null;
         }
-        return null;
     }
 
-    public TransactionState getTransactionStateByCallbackId(long callbackId) {
-        readLock();
+    public TransactionState getTransactionStateByCallbackId(long dbId, long callbackId) {
         try {
-            for (TransactionState txn : idToTransactionState.values()) {
-                if (txn.getCallbackId() == callbackId) {
-                    return txn;
-                }
-            }
-        } finally {
-            readUnlock();
+            DatabaseTransactionMgr dbTransactionMgr = getDatabaseTransactioMgr(dbId);
+            return dbTransactionMgr.getTransactionStateByCallbackId(callbackId);
+        } catch (AnalysisException e) {
+            LOG.warn("Get transaction by callbackId failed", e);
+            return null;
         }
-        return null;
     }
 
-    public List<Long> getTransactionIdByCoordinateBe(String coordinateHost, int limit) {
-        ArrayList<Long> txnIds = new ArrayList<>();
-        readLock();
-        try {
-            idToTransactionState.values().stream()
-                    .filter(t -> (t.getCoordinator().sourceType == TransactionState.TxnSourceType.BE
-                            && t.getCoordinator().ip.equals(coordinateHost)
-                            && (!t.getTransactionStatus().isFinalStatus())))
-                    .limit(limit)
-                    .forEach(t -> txnIds.add(t.getTransactionId()));
-        } finally {
-            readUnlock();
+    public List<Pair<Long, Long>> getTransactionIdByCoordinateBe(String coordinateHost, int limit) {
+        ArrayList<Pair<Long, Long>> txnInfos = new ArrayList<>();
+        for (DatabaseTransactionMgr databaseTransactionMgr : dbIdToDatabaseTransactionMgrs.values()) {
+            txnInfos.addAll(databaseTransactionMgr.getTransactionIdByCoordinateBe(coordinateHost, limit));
+            if (txnInfos.size() > limit) {
+                break;
+            }
         }
-        return txnIds;
+        return txnInfos.size() > limit ? new ArrayList<>(txnInfos.subList(0, limit)) : txnInfos;
```

Comment:
```suggestion
        return txnInfos.size() > limit ? txnInfos.subList(0, limit) : txnInfos;
```

Reply:
@morningman  subList is a view of origin list, so is it a good way to directly return sublist?

Reply:
I think it's OK, cause no one will modify that list.